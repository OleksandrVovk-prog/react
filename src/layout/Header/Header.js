import { Link } from 'react-router-dom';
import { useTranslation } from 'react-i18next';
import Title from '../../components/Title/Title';
import { Context as PageContext } from '../../context/layout/Page';
import styles from './sass/Header.module.scss';

export default function Header() {
  const { i18n, t } = useTranslation();

  return (
    <PageContext.Consumer>
      {(config) => (
        <header className={styles.appHeader}>
          <Title text={t('welcome')} />
          <Link to="/" className={styles.headerLink}>
            {t('home')}
          </Link>
          <Link to="/about" className={styles.headerLink}>
            {t('about')}
          </Link>
          <Link to="/lazy" className={styles.headerLink}>
            {t('lazy')}
          </Link>
          <Link to="/memo" className={styles.headerLink}>
            {t('memo')}
          </Link>
          <Link to="/context" className={styles.headerLink}>
            {t('context')}
          </Link>
          <button
            className={styles.toggleLanguage}
            type="button"
            onClick={config.changeLanguage}
            data-testid="toggle-language-button"
          >
            {i18n.language}
          </button>
        </header>
      )}
    </PageContext.Consumer>
  );
}
