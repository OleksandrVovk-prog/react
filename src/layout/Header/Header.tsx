import { Link } from 'react-router-dom';
import { useTranslation } from 'react-i18next';
import { useAppDispatch, useAppSelector } from '../../store/hooks/useApp';
import { toggleLocale } from '../../store/slices/translates/slice';
import { selectLocale } from '../../store/slices/translates/selectors';
import Title from '../../components/Title/Title';
import { selectUserId } from '../../store/slices/auth/selectors';
import authApi, { useFetchUserQuery } from '../../store/slices/auth/apis/auth';

import styles from './sass/Header.module.scss';

function Header(): JSX.Element {
  const { t } = useTranslation();
  const dispatch = useAppDispatch();
  const id = useAppSelector(selectUserId);
  const locale = useAppSelector(selectLocale);
  const { data } = useFetchUserQuery(id, { skip: !id });
  return (
    <header className={styles.header}>
      <Title text={t('common.welcomeMessage')} />
      <nav className={styles.headerNav}>
        <Link to="/" className={styles.headerNavLink}>
          { t('home.pageTitle') }
        </Link>
        <Link to="/about" className={styles.headerNavLink}>
          { t('about.pageTitle') }
        </Link>
      </nav>
      <div className={styles.toggleLanguage}>
        {data?.id ? (
          <Link
            to="/login"
            onClick={() => {
              dispatch(authApi.util.resetApiState());
            }}
            className={styles.headerNavLink}
          >
            { `${t('login.logout')}(${data.firstName})` }
          </Link>
        ) : (
          <Link
            to="/login"
            className={styles.headerNavLink}
          >
            { t('login.pageTitle') }
          </Link>
        )}
        <button
          data-testid="toggle-language-button"
          type="button"
          onClick={() => dispatch(toggleLocale())}
        >
          {locale.toUpperCase()}
        </button>
      </div>
    </header>
  );
}

export default Header;
