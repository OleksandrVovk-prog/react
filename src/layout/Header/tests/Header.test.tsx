import { screen } from '@testing-library/react';
import render from '../../../../__tests__/test-utils';

import Header from '../Header';

describe('<Header />', () => {
  test('should render component', () => {
    const { asFragment } = render(<Header />);
    expect(asFragment()).toMatchSnapshot();
  });
  test('should contain needed elements', () => {
    render(<Header />);
    const title = screen.getByText('Try to build something great');
    const listLinks = screen.getAllByRole('link');
    const button = screen.getByTestId('toggle-language-button');
    const loginLink = screen.getByText('Login');
    expect(title).toBeInTheDocument();
    expect(loginLink).toBeInTheDocument();
    expect(button).toBeInTheDocument();
    expect(listLinks).toHaveLength(3);
  });
});
