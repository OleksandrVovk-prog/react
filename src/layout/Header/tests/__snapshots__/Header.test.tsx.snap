// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> should render component 1`] = `
<DocumentFragment>
  <header
    class="header"
    data-testid="header"
  >
    <h2
      class="appSubtitle"
      data-testid="subtitle"
    >
      Try to build something great
    </h2>
    <nav
      class="headerNav"
    >
      <a
        class="headerNavLink"
        href="/"
      >
        Home
      </a>
      <a
        class="headerNavLink"
        href="/about"
      >
        About
      </a>
    </nav>
    <div
      class="toggleLanguage"
    >
      <span>
        <a
          class="linkButton"
          href="/login"
        >
          Login
        </a>
      </span>
      <button
        data-testid="toggle-language-button"
        type="button"
      >
        EN
      </button>
    </div>
  </header>
</DocumentFragment>
`;
